hal/src/common/rtc.rs

```diff
@@ -0,0 +1,305 @@
+        if enable {
+            self.mode0_ctrla().modify(|_, w| w.enable().set_bit());
+        } else {
+            self.mode0_ctrla().modify(|_, w| w.enable().clear_bit());
+        }
+        self.sync();
+    }
+
+    // --- clock functions
+
+    /// Configures the peripheral for clock/calendar mode. Requires the source
+    /// clock to be running at 1024 Hz.
+    pub fn clock_mode(&mut self) {
+        // The max divisor is 1024, so to get 1 Hz, we need a 1024 Hz source.
+        assert_eq!(self.rtc_clock_freq.0, 1024_u32, "RTC clk not 1024 Hz!");
+
+        self.mode2_ctrla().modify(|_, w| {
+            w.mode().clock() // enable mode2 (clock)
+            .clkrep().clear_bit()
+            .matchclr().clear_bit()
+            .prescaler().div1024() // 1.024 kHz / 1024 = 1Hz
+        });
+        self.sync();
+
+        // enable clock sync on SAMx5x
+        #[cfg(feature = "min-samd51g")]
+        {
+            self.mode2_ctrla().modify(|_, w| {
+                w.clocksync().set_bit() // synchronize the CLOCK register
+            });
+
+            self.sync();
+        }
+
+        self.enable(true);
+    }
+
+    /// Returns the current clock/calendar value.
+    pub fn current_time(&self) -> Datetime {
+        // synchronize this read on SAMD11/21. SAMx5x is automatically synchronized
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            self.mode2().readreq.modify(|_, w| w.rcont().set_bit());
+            self.sync();
+        }
+        self.mode2().clock.read().into()
+    }
+
+    /// Updates the current clock/calendar value.
+    pub fn set_time(&mut self, time: Datetime) {
+        self.mode2().clock.write(|w| unsafe {
+            w.second()
+                .bits(time.seconds)
+                .minute()
+                .bits(time.minutes)
+                .hour()
+                .bits(time.hours)
+                .day()
+                .bits(time.day)
+                .month()
+                .bits(time.month)
+                .year()
+                .bits(time.year)
+        });
+        self.sync();
+    }
+}
+
+// --- Timer / Counter Functionality
+
+impl Periodic for Rtc {}
+impl CountDown for Rtc {
+    type Time = Nanoseconds;
+
+    fn start<T>(&mut self, timeout: T)
+    where
+        T: Into<Self::Time>,
+    {
+        let params = TimerParams::new_us(timeout, self.rtc_clock_freq.0);
+        let divider = params.divider;
+        let cycles = params.cycles;
+
+        // Disable the timer while we reconfigure it
+        self.enable(false);
+
+        // 32 bit counter mode plz
+        self.mode0_ctrla().modify(|_, w| w.mode().count32());
+
+        // Now that we have a clock routed to the peripheral, we
+        // can ask it to perform a reset.
+        self.reset();
+
+        // the SVD erroneously marks swrst as write-only, so we
+        // need to manually read the bit here
+        while self.mode0_ctrla().read().bits() & 1 != 0 {}
```

Comment:
You might see if @ianrrees can help you push a change upstream to Microchip, to correct the SVD. It looks like there might be other areas in the RTC that could use changes or corrections as well.

Reply:
That'd be great! This seems to be an issue on the RTC & timer peripherals

Reply:
Yep, I'm planning on assembling a list of fixes on Monday, will add this one.